src/useAsnyc.js

```diff
@@ -4,42 +4,107 @@ import * as React from 'react';
 import usePromise from './utils/usePromise';
 
 class AsyncStore {
-  constructor(result, error, state) {
-    this.result = result;
+  constructor(result, error, state, latestValue) {
+    this._result = result;
     this._error = error;
-    this.state = state;
+    this._state = state;
+    this._latestValue = latestValue;
+  }
+
+  hasValue() {
+    return this._latestValue != null;
+  }
+
+  hasError() {
+    return this._error != null;
```

Comment:
I think we're removing .error/.hasError since we changed it to not preserve the error across state changes. We could remove those and rename isFalure/failure to isError/error. I'm not sure which naming is better.